kernel/comps/virtio/src/device/network/device.rs

```diff
@@ -185,6 +186,19 @@ impl NetworkDevice {
         debug_assert!(self.tx_buffers[token as usize].is_none());
         self.tx_buffers[token as usize] = Some(tx_buffer);
 
+        self.free_processed_tx_buffers();
+
+        // If the send queue is not full, we can free the send buffers during the next sending process.
+        // Therefore, there is no need to free the used buffers in the IRQ handlers.
+        // This allows us to temporarily disable the send queue interrupt.
+        // Conversely, if the send queue is full, the send queue interrupt should remain enabled
+        // to free the send buffers as quickly as possible.
+        if !self.can_send() {
+            self.send_queue.enable_callback();
+        } else {
+            self.send_queue.disable_callback();
+        }
```

Comment:
It seems that this mechanism also makes sense for the receive path, e.g. if there is something left in the receive queue, we can disable the receive IRQ/callback.

However, currently we poll the interfaces directly in the IRQ handler, and the interface polling will always consume all packets from the device, so we will never leave the receive queue in a non-empty state. So doing the same thing for the receive path becomes unnecessary, at least in our current codebase, right?

Reply:
> It seems that this mechanism also makes sense for the receive path, e.g. if there is something left in the receive queue, we can disable the receive IRQ/callback.

Yes, I agree that the optimization makes sense. If there are more packets remaining in the receive queue, we can enter a busy polling mode until all packets are received, allowing us to temporarily disable interrupts. However, if I implement this mechanism, there wouldn't be any observable performance benefits.

> However, currently we poll the interfaces directly in the IRQ handler, and the interface polling will always consume all packets from the device, so we will never leave the receive queue in a non-empty state. So doing the same thing for the receive path becomes unnecessary, at least in our current codebase, right?

I think you are correct. As it stands, we are effectively performing busy polling in the IRQ handler, so the optimization does not make sense. If we were to move the polling logic to another context, such as the softirq context, and limit the IRQ handler to only transferring packets from the virtual queue, then the optimization could be adopted.